pkg/dns/aws/dns.go

```diff
@@ -49,120 +47,155 @@ const (
-	switch region {
-	case endpoints.CnNorth1RegionID, endpoints.CnNorthwest1RegionID:
-		r53Config = r53Config.WithRegion(endpoints.CnNorthwest1RegionID).WithEndpoint(chinaRoute53Endpoint)
-	case endpoints.UsGovEast1RegionID, endpoints.UsGovWest1RegionID:
-		// Route53 for GovCloud uses the "us-gov-west-1" region id:
-		// https://docs.aws.amazon.com/govcloud-us/latest/UserGuide/using-govcloud-endpoints.html
-		r53Config = r53Config.WithRegion(endpoints.UsGovWest1RegionID)
-	case endpoints.UsIsoEast1RegionID:
-		// Do not override the region in C2s
-		r53Config = r53Config.WithRegion(region)
-	default:
-		// Use us-east-1 for Route 53 in AWS Regions other than China or GovCloud Regions.
-		// See https://docs.aws.amazon.com/general/latest/gr/r53.html for details.
-		r53Config = r53Config.WithRegion(endpoints.UsEast1RegionID)
-	}
-
-	p := &Provider{
-		route53: &InstrumentedRoute53{route53.New(sess, r53Config)},
-		config:  config,
-		logger:  log.Log.WithName("aws-route53").WithValues("region", r53Config.Region),
+		return nil, fmt.Errorf("unable to create aws session: %s", err)
+	}
+	if route53Config.Region != "" {
+		sess.Config.WithRegion(route53Config.Region)
+	}
+
+	p := &Route53DNSProvider{
+		client: &InstrumentedRoute53{route53.New(sess, config)},
+		logger: log.Log.WithName("aws-route53").WithValues("region", config.Region),
 	}
+
 	if err := validateServiceEndpoints(p); err != nil {
 		return nil, fmt.Errorf("failed to validate AWS provider service endpoints: %v", err)
 	}
-	//if p.healthCheckReconciler == nil {
-	//	p.healthCheckReconciler = newRoute53HealthCheckReconciler(p.route53, p.logger)
-	//}
 
 	return p, nil
 }
 
-// validateServiceEndpoints validates that provider clients can communicate with
-// associated API endpoints by having each client make a list/describe/get call.
-func validateServiceEndpoints(provider *Provider) error {
-	var errs []error
-	zoneInput := route53.ListHostedZonesInput{MaxItems: aws.String("1")}
-	if _, err := provider.route53.ListHostedZones(&zoneInput); err != nil {
-		errs = append(errs, fmt.Errorf("failed to list route53 hosted zones: %v", err))
-	}
-	return kerrors.NewAggregate(errs)
-}
-
 type action string
 
 const (
 	upsertAction action = "UPSERT"
 	deleteAction action = "DELETE"
 )
 
-func (p *Provider) Ensure(record *v1.DNSRecord, zone v1.DNSZone) error {
-	return p.change(record, zone, upsertAction)
+func (p *Route53DNSProvider) Ensure(record *v1alpha1.DNSRecord, managedZone *v1alpha1.ManagedZone) error {
+	return p.change(record, managedZone, upsertAction)
+}
+
+func (p *Route53DNSProvider) Delete(record *v1alpha1.DNSRecord, managedZone *v1alpha1.ManagedZone) error {
+	return p.change(record, managedZone, deleteAction)
+}
+
+func (p *Route53DNSProvider) EnsureManagedZone(zone *v1alpha1.ManagedZone) (ManagedZoneOutput, error) {
+	var zoneID string
+	if zone.Spec.ID != "" {
+		zoneID = zone.Spec.ID
+	} else {
+		zoneID = zone.Status.ID
```

Comment:
Is there an edge case where the ID is not set in either?

Reply:
If `zone.Status.ID` isn't set then we need to create one, and after we create it updates the status with the ID.